### ISSUE TITLE
ms/ms with full-scan ms2

### ISSUE THREAD

**Issue by chubukov** [ElucidataInc/ElMaven#170]
reported by/with @pirman12 

There's an extremely common ms/ms method type that Maven almost handles, but doesn't really do well on: selecting on ms1 and then doing full scan on ms2. It really seems to particularly fail if you have more than one ms1 m/z that you're monitoring.

I uploaded a few sample data files to dropbox.

Ideally you should be able to specify precursorMz and productMz, and it should find the relevant scan and then find the relevant ion (based on ppm)*. It looks like right now as long as you specify the parent of the first scan, it basically works -- in the sample compound database I gave, it finds test1 but not test2. 

But the second scan is recognized and stored correctly. For instance, if I open the SRM list widget and click on the 209.19 parent scan, I get the TIC (ok) but also if I open the spectrum widget, it gets the right spectrum and correctly annotates with with precursorMz=209.19. (however, clicking on one of the spectrum bars does not do the right thing). Moreoever, if I open the fragmentation widget (not in the sidebar, but it is in the widgets menu), it definitely recognizes the parent scans, if I put in the parent mass into the main m/z window. Funnily, for 203.15, it also recognizes it from the SRM list window, but for 209.19 it does not. 

So basically it looks like
 - data is being read in more or less correctly
 - there is a problem pulling the right EIC for the second "transition"
 - clicking on the spectrum bars does not have the right effect
 - variety of other tools developed for full scan MS1 data don't really do the right thing: e.g. if I bookmark the "test1" peak I get the parent mass for the "expectedMz" but the daughter mass for the "observedMz". These columns (and also the columns in sample widget should probably just be different for MS/MS data, as suggested in #67)

*probably the Q1 amu setting should be used to find the right parent and then the ppm setting to find the daughter ion.

**Comment by chubukov** [comment]
I'm putting this as high priority, but it doesn't need to be a hotfix -- we have some workarounds, and besides, it may not be trivial to get it right. Let's look at it carefully in the next sprint. But I do think that fixing stuff like this is much higher priority than new features like loading libraries.

**Comment by chubukov** [comment]
Also, if I put in a potential daughter ion mass into the main m/z window, then it looks to me like it is extracting the EIC from the 203.15 parent scan, so not even the full scan. I'm not sure what the right behavior is here, but i thought by default it would look in all the scans?

**Comment by pirman12** [comment]
It would be great if for MS/MS scans an EIC query would search spectra at the MS/MS level and we could chose which MS/MS scan we were interested in via the "show SRM list" button.  If I had 3 scan functions: 1) full scan MS 2) MS/MS of m/z 300 and 3) MS/MS of 400; I input an XIC in the top right or chose a metabolite with the correct m/z value that I'm interested in and the XIC would displayed based on which SRM list I had highlighted.  
Another option would be that the maven source file contains this information like it does for MRM methods on a QExactive.  If I know the parent mass is 300 and the daughter is 200, I could just import that information and then highlight the correct SRM to generate the XIC.

**Comment by chubukov** [comment]
@pirman12 How about if we simply have an MS1 window and an MS2 window where there is currently the m/z window on the top right? The functionality is basically there already (there is a Q1 amu window and a Q3 amu window in the SRM widget). The MS2 window could be disabled if there are no MS2 scans.

That seems simplest to me. And the Maven source files would work the same way, just put in precursorMz and productMz and then specify the ppm or amu resolution in the top right boxes.

**Comment by chubukov** [comment]
@sahil21 btw, I discussed this with our group a little bit, and I think the suggestion outlined in the last comment would be sufficient for 80-90% of our needs. And at least to me, it seems like a very intuitive interface.

**Comment by shubhra-agrawal** [comment]
@chubukov  @pirman12  Instead of adding another window for search, we have allowed users to enter hyphen-separated precursor and product m/z values to search for MS2 scans. For example: 121-102 

Currently, the ppm window only works for LCMS data or single value search though there are plans to add this functionality for MRM search at a later point in time. It should be a much cleaner fix once we have integrated MS/MS properly in the code base. For the time being, resolution can be adjusted from the Q1 and Q3 boxes on the SRM widget for MRM data.

**Comment by chubukov** [comment]
@shubhra-agrawal I have tested this feature a little bit. I am only able to get EICs extracted from the first transition in the list. E.g. if have a file with the following scan types

1. 250 -> [50-250]
2. 200 -> [50-250]
3. 150 -> [50-250]
4. MS1 [50-500]

I only get proper EICs for the 250-> X, and not for 200 -> X

**Comment by chubukov** [comment]
I take back my last comment (leaving it up for history). It seems pretty random which EICs work and which ones don't -- sometimes I do see the correct ones for the other scan types. But for some daughters I don't get the EIC pulled out even from the first scan type.

**Comment by shubhra-agrawal** [comment]
@chubukov we were looking into this issue and noticed the merged EIC for 209->[50-230]:
![mergedeic](https://user-images.githubusercontent.com/28925846/31485923-3a4536f4-af53-11e7-8f38-68848607562e.png)

Even for MS1 peaks, the merged EIC looks very odd and broken. Is this true for all full-scan MS2 datasets or is this a dataset-specific issue we're seeing here?

**Comment by sp-eldata** [comment]
Very artsy - the first visualization :)

**Comment by chubukov** [comment]
@shubhra-agrawal a little difficult for me to judge what's happening, but there shouldn't be anything particularly degenerate about this data. I'm guessing the merged EIC weirdness is because the scan type isn't getting restricted properly? Not sure, but it should be easy to check -- just pull out the EICs and ask what the merged EIC should look like. (it's just a smoothed average).

**Comment by shubhra-agrawal** [comment]
@chubukov While I could pull EICs for all precursor-product m/z combinations in the above data set, I do get inconsistent results with the data you shared with us yesterday. I'll let you know when I figure something out.

**Comment by shubhra-agrawal** [comment]
@chubukov I realized that was because of zooming issues. Sometimes when I use the text-search feature, I have to click on zoom-out to centre the peaks. I'll look into that. Apart from that, I can find EICs for all precursor-product combinations using the text-search feature, given the Q1 and Q3 amu values in the SRM List widget are not set to 0. 

I printed out the scans for some very choppy peaks that have weird merged EICs. Turns out there are only 4-5 data points per sample for those so that's to be expected.

**Comment by chubukov** [comment]
@shubhra-agrawal Yes, the zoom is not set well by default, but everything I've mentioned has been after zooming out completely. Usually I just get the "Empty EIC" grey text displayed.

**Comment by shubhra-agrawal** [comment]
@chubukov Just asking so we're on the same page, did you check if the amu settings were non-zero? If that was non-zero then I'd have to test it out on a windows system and see if there's a difference.

**Comment by chubukov** [comment]
@shubhra-agrawal Yes, 0.5 and 0.5 amu. I guess it's possible that it's a windows/linux issue but that would not be my guess. Why don't we go over specific cases by email first.

**Comment by chubukov** [comment]
@shubhra-agrawal Actually I think I know what's going on. I think mostly it's that the spectrum widget is not showing what I think it is. There may be some other issues as well.

**Comment by shubhra-agrawal** [comment]
@chubukov 
`I think mostly it's that the spectrum widget is not showing what I think it is.` 
Can you elaborate? I think it would be best if you could share some screenshots over email where things don't seem to match up with what you would expect.

**Comment by chubukov** [comment]
@shubhra-agrawal sure I'll put something together. I think this is not the only issue. But if you select a function from the SRM list widget, then click on a peak in the EIC, the spectrum widget seems to populate with the product ions for that precursor (based on the text at the top of the widget). But when you do "reset zoom" I think it's actually showing all the ions for all the scans, even though the text is the same.

**Comment by shubhra-agrawal** [comment]
@chubukov As you pointed out in the email, El-MAVEN was plotting EICs using only the scans where m/z of the highest intensity ion (basepeakMz) matched the provided product m/z. Naturally, it led to loss of data points which is why we were getting broken mergedEICs like I posted above. 
I have changed the getEIC function to look at every fragment m/z for the correct precursor scans to find the product m/z. This is consistent with how getEIC works for LCMS data.
I have shared the screenshots with you on email of the fixed version. This has also improved automated peak detection for full-scan ms2 data. Intensities for MRM data will be a little more accurate now as well. We have merged this fix into develop. Let me know once you have tested it out.
